drivers/pinctrl/pinctrl_esp32.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2022 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <drivers/gpio.h>
+#include <drivers/pinctrl.h>
+#include <drivers/pinctrl/pinctrl_esp32_common.h>
+
+/**
+ * @brief Array containing pointers to each GPIO port.
+ *
+ * Entries will be NULL if the GPIO port is not enabled.
+ */
+static const struct device * const gpio_ports[] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio0)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio1)),
+};
+
+/** Number of GPIO ports. */
+static const size_t gpio_ports_cnt = ARRAY_SIZE(gpio_ports);
+
+static int esp32_pin_configure(const uint32_t pin_mux, const uint32_t pin_cfg)
+{
+	const struct device *port_device;
+	uint32_t pin_num = ESP32_PIN_NUM(pin_mux);
+	uint32_t sig_in = ESP32_PIN_SIGI(pin_mux);
+	uint32_t sig_out = ESP32_PIN_SIGO(pin_mux);
+	int flags = 0;
+
+	if (ESP32_PORT_IDX(pin_num) >= gpio_ports_cnt) {
+		return -EINVAL;
+	}
+
+	port_device = gpio_ports[ESP32_PORT_IDX(pin_num)];
+
+	if ((port_device == NULL) || (!device_is_ready(port_device))) {
+		return -ENODEV;
+	}
```

Comment:
please don't mix GPIO/pinctrl. pinctrl drivers need to be independent, and, if there's ever a relation it'll likely be GPIO depending on pinctrl (since in most cases it's a subset). Same discussion happened for the Microchip case, see https://github.com/zephyrproject-rtos/zephyr/pull/43282

Reply:
@gmarull thanks for pointing that out. I have decoupled the pinctrl driver from the GPIO driver, removing dependency between them.

Though you will see some gpio-prefixed naming are kept here and there, it is due to hal_espressif's own conventions.

Reply:
I still see the gpio based implementation on the first commit, can you please squash//fixup?

Reply:
Squashed/Fixed.